letsencrypt/models/letsencrypt.py

```diff
@@ -2,167 +2,223 @@
+        client = self._create_client(account_key)
+        new_reg = NewRegistration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        client.new_account(new_reg)
+        csr = self._make_csr(account_key, domain_key_file, domain)
+        authzr = client.new_order(csr)
+        # write the file with the challenge
+        challenge_dir = _get_challenge_dir()
+        if not isdir(challenge_dir):
+            makedirs(challenge_dir)
+        auth_responded = False
+        for authorizations in authzr.authorizations:
+            for challenge in sorted(
+                    authorizations.body.challenges, key=lambda x: x != 'http'):
+                if challenge.chall.typ == TYPE_CHALLENGE_HTTP:
+                    self._respond_challenge_http(challenge, account_key)
+                    client.answer_challenge(
+                        challenge,
+                        challenges.HTTP01Response())
+                    auth_responded = True
+                    break
+                elif challenge.chall.typ == TYPE_CHALLENGE_DNS:
+                    self._respond_challenge_dns(challenge, domain)
+                    client.answer_challenge(
+                        challenge, challenges.DNSResponse())
+                    auth_responded = True
+                    break
+        if not auth_responded:
+            raise exceptions.ValidationError(
+                _('Could not respond to letsencrypt challenges.'))
+        # let them know we are done and they should check
+        order_resource = client.poll_and_finalize(authzr)
+        with open(join(_get_data_dir(), '%s.crt' % domain), 'w') as crt:
+            crt.write(order_resource.fullchain_pem)
+            _logger.info('SUCCESS: Certificate saved :%s', crt.name)
+            ir_config_parameter = self.env['ir.config_parameter']
+            reload_cmd = ir_config_parameter.get_param(
+                'letsencrypt_reload_command', False)
+            if reload_cmd:
+                self._call_cmdline(['sh', '-c', reload_cmd])
+            else:
+                _logger.warning("No reload command defined.")
+
+    def _create_client(self, account_key):
+        net = client.ClientNetwork(account_key)
+        if bool(self.env['ir.config_parameter'].get_param('letsencrypt_demo')):
+            directory_url = V2_STAGING_DIRECTORY_URL
+        else:
+            directory_url = V2_DIRECTORY_URL
+        directory_json = requests.get(directory_url).json()
+        return client.ClientV2(directory_json, net)
+
+    def _make_csr(self, account_key, domain_key_file, domain):
+        parameter_model = self.env['ir.config_parameter']
+        altnames = parameter_model.search([
+            ('key', '=', 'letsencrypt_altnames')],
+        ).value
+        if altnames:
+            altnames = altnames.split(',')
+            valid_domains = altnames + [domain]
+        else:
+            valid_domains = [domain]
+        _logger.info(
```

Comment:
My problem with this is I could not find a standard way of extracting the domain from these urls. While `urlparse` can give you the whole domain, there does not seem to be a standard way of extracting the second level domain. Ex. www.example.eu, www.example.co.uk etc. Any ideas?

Reply:
there aren't any [URLs](https://en.wikipedia.org/wiki/URL) at this stage, it's just domains.

Think of an algorithm yourself. Up to now I thought the asterisk can only be the first domain component, but according to https://www.ietf.org/rfc/rfc2818.txt it can also be part of a component. My first idea would be to take all domains that contain an asterisk, replace the asterisk with `[^.*]*`, prepend the string with `^` and append `$`. Feed this to `re.compile` and throw away all domains that match this. And read up on regexes why this should work (untested) in case you're not acquainted to them.

Studying this a little bit more, I see that `*.therp.nl` and `therp.nl` actually are different things from the point of view of wildcard certificates, so we do indeed need to have two requests there and what I wrote above is wrong. Still we want to reduce the list to save unnecessary challenges.

But make it so that when there's a DNS provider configured, you prefer DNS. This way, we support non-publicly accessible odoo instances which create their SSL certificates via DNS.

Reply:
> there aren't any URLs at this stage, it's just domains.
I was under the impression that the users will input urls in the field since:
> The name MUST include both a scheme (e.g., "http" or "ftp") and a scheme-specific-part.
Also,
>  Still we want to reduce the list to save unnecessary challenges.
I did not see on the rfc what happens in the case we have a wildcard and a domain on the same level, but according to this https://github.com/letsencrypt/boulder/blob/b4c13ce6c7d92ce6cd0de2fc8f78b7f58b928d95/ra/ra.go#L1945 we will get an error.
Besides this though, the altnames can also accept other things than domains (IP addresses, emails etc as written here https://tools.ietf.org/html/rfc5280#section-4.2.1.6)
So my argument for this would be to just split the elements and send them as they are.

In any case, If I am to implement this, this is how I would do it (and let me know if its correct):
```
if len(domains)> 1:
  for domain in filter(lambda x: '*.' in x, domains):
    ind = domain.index('*.') + 2 
    to_delete = filter(lambda x: x[ind:] == domain, domains)
    for to_delete_item in to_delete:
       del domains[to_delete_item]
return domains
```

Would that do?

Reply:
try it

And please turn around the order of trying https vs dns verification, so that if we have a dns provider, we do dns verification